### DISCUSSION TITLE
Custom jinja filters - JINJA2_FILTERS

### DISCUSSION THREAD

**Discussion by Kani999** [netbox-community/netbox#11759] [Q&A] [answered]
Hello, 

I'm trying to use JINJA2_FILTERS.

I've created:
`configuration.py`
```python
...
# JINJA2_FILTERS
def to_czech_crown(number):
    if number:
        return '{:,}'.format(number).replace(',', ' ') + " Kč"
    else:
        return '---'

JINJA2_FILTERS = {'to_czech_crown': to_czech_crown}
...
```

Now I want to use the filter in my plugin:
`invoices.html`
```
   <td>{{ invoice.price|to_czech_crown }}</td>
```

But error occurs:
![image](https://user-images.githubusercontent.com/17337078/219305881-e4920ef4-9413-4b85-8c68-b8ba7be57cb2.png)

I found out, that the `render` method in my `template_content.py`:
```python
        return self.render(
            'inventory_monitor/invoices_include.html',
            extra_context={
                'object': object,
                'invoices': invoices,
            }
        )
```
does not call the `render_jinja2` method at `utilities/utils.py`:
```python
def render_jinja2(template_code, context):
    """
    Render a Jinja2 template with the provided context. Return the rendered content.
    """
    import ipdb; ipdb.set_trace()

    
    environment = SandboxedEnvironment()
    environment.filters.update(get_config().JINJA2_FILTERS)
    
    
    return environment.from_string(source=template_code).render(**context)
```

How can I add custom jinja filter to be in use in template_content rendering?

---
Edit:

In the settings, I can see my custom filter:
```
import netbox.settings as netbox_settings
netbox_settings.JINJA2_FILTERS
>>> {'to_czech_crown': <function to_czech_crown at 0x7fa98cf1f010>}
```

**Comment by Kani999** [comment]
Adding:

```python
@register.filter()
def to_czech_crown(number):
    if number:
        return '{:,}'.format(number).replace(',', ' ') + " Kč"
    else:
        return '---'
```

To `netbox/utilities/templatetags/builtins/filters.py` works, but I don't want to add the function to the netbox core.

**Reply by Kani999** [reply]
Adding to my plugin `views.py`:

```
from utilities.templatetags.builtins.filters import register

@register.filter()
def to_czech_crown(number):
    if number:
        return '{:,}'.format(number).replace(',', ' ') + " Kč"
    else:
        return '---'
```

solves the problem. But I don't think this is the best way.

**Reply by kkthxbye-code** [reply]
The feature is intended to register filters for user created templates, like webhook body template, export templates and custom links. These places use jinja2 as a template engine. The normal templates and templates used in plugins use the django templating engine and you should register templatetags for these yourself in your plugin.

**Reply by Kani999** [reply]
Okay thank you. I did not found this info in the docs.

**Comment by Kani999** [comment, marked as answer]
@kkthxbye-code answer

The feature is intended to register filters for user created templates, like webhook body template, export templates and custom links. These places use jinja2 as a template engine. The normal templates and templates used in plugins use the django templating engine and you should register templatetags for these yourself in your plugin.

**Comment by zhabchenko** [comment]
Hi! 
i adding my custom filter in netbox/utilities/templatetags/builtins/filters.py, but but it still won't load, and even standard filters from this repository do not work.
![image](https://user-images.githubusercontent.com/98376538/235852570-8b036dd5-875e-40c4-a979-94d8cf216d8c.png)

**Reply by kkthxbye-code** [reply]
Modifying the core netbox code without understanding it, and adding a django template filter instead of a jinja2 filter will obviously not work. Please follow the documentation:

https://docs.netbox.dev/en/stable/configuration/system/#jinja2_filters

**Reply by zhabchenko** [reply]
Yes, it's worked.  Thanks!
